### PR TITLE
[WIP] Setup electron-builder

### PR DESCRIPTION
**Pull request by jam-world**
Setup the pipeline to build distributable app with electron-builder.

### CODE REVIEW

File: file-browser/package.json [hunk 1]
```diff
@@ -22,17 +22,19 @@
   },
   "private": true,
   "dependencies": {
+    "electron-is-dev": "^0.2.0",
+    "eywa": "git+https://github.com/escherpad/eywa.git",
```
Comment (episodeyang) [inline file-browser/package.json]: @jam-world On a side note, we might want to change eywa's name and publish it somewhere.

File: file-browser/package.json [hunk 2]
```diff
@@ -22,17 +22,19 @@
   },
   "private": true,
   "dependencies": {
+    "electron-is-dev": "^0.2.0",
```
Comment (episodeyang) [inline file-browser/package.json]: what does this mean?

File: file-browser/package.json [hunk 3]
```diff
@@ -22,17 +22,19 @@
...
+    "eywa": "git+https://github.com/escherpad/eywa.git",
     "layout-components": "^2.7.1",
     "luna": "^1.5.0",
-    "luna-saga": "^3.1.0",
+    "luna-saga": "^3.0.1",
+    "npm": "^5.1.0",
```
Comment (episodeyang) [inline file-browser/package.json]: this is strange. Do we need this?

File: file-browser/package.json [hunk 4]
```diff
@@ -22,17 +22,19 @@
   },
   "private": true,
   "dependencies": {
+    "electron-is-dev": "^0.2.0",
+    "eywa": "git+https://github.com/escherpad/eywa.git",
     "layout-components": "^2.7.1",
     "luna": "^1.5.0",
-    "luna-saga": "^3.1.0",
+    "luna-saga": "^3.0.1",
```
Comment (episodeyang) [inline file-browser/package.json]: this seems wrong. Why don't we stay on 3.1.0? would you please double check this? Thanks!

- [ ] make sure that @3.1.0 works.
- [ ] remote this edit.

File: file-browser/package.json [hunk 5]
```diff
@@ -22,17 +22,19 @@
...
+    "npm": "^5.1.0",
     "react": "^15.4.2",
     "react-dom": "^15.4.2",
     "rxjs": "^5.3.0",
     "uuid": "^3.0.1"
   },
   "devDependencies": {
     "concurrently": "^3.4.0",
-    "eywa": "git://github.com/escherpad/eywa.git",
```
Comment (episodeyang) [inline file-browser/package.json]: remove this. eywa should be in `dependencies` only. *not* in `devDepe..`.

File: file-browser/package.json [hunk 6]
```diff
@@ -47,6 +49,20 @@
     "build": "react-scripts build",
     "test": "react-scripts test --env=jsdom",
     "eject": "echo 'Warning: not reversible!! execute: \"react-scripts eject\"'",
-    "electron": "electron ."
+    "electron": "electron .",
+    "electron-dev": "concurrently \"BROWSER=none yarn start\" \"wait-on http://localhost:3000 && electron .\"",
+    "electron-pack": "build --em.main=build/electron.js",
+    "preelectron-pack": "yarn build"
```
Comment (episodeyang) [inline file-browser/package.json]: awesome. Good work!!

File: file-browser/public/electron.js [hunk 7]
```diff
@@ -9,14 +10,8 @@ function createWindow() {
...
-    // and load the index.html of the app.
-    // from https://medium.freecodecamp.com/building-an-electron-application-with-create-react-app-97945861647c
-    const startUrl = process.env.ELECTRON_START_URL || url.format({
-            pathname: path.join(__dirname, 'build/index.html'),
-            protocol: 'file:',
-            slashes: true
-        });
-    win.loadURL(startUrl);
+    // load the url
+    win.loadURL(isDev ? 'http://localhost:3000' : `file://${path.join(__dirname, '../build/index.html')}`);
```
Comment (episodeyang) [inline file-browser/public/electron.js]: This edit seems wrong. Can you double check?
- [ ] make the edit as minimal as possible.

File: file-browser/src/model/dataSchema.js [hunk 8]
```diff
@@ -4,6 +4,7 @@ declare type AccountT = {
     $key: string, // locally generated uuid, used internally in our app
     client_id: string, // OAuth client_id
     access_token: string, // OAuth token
+    backend: string,
```
Comment (episodeyang) [inline file-browser/src/model/dataSchema.js]: remove this. This belongs to a different PR. 

Also it shouldn't be `backend`, it should be `service`?

File: file-browser/src/model/model.js [hunk 9]
```diff
@@ -2,24 +2,20 @@
...
 
 const DEFAULT_ACCOUNTS_STATE = {
-    "dropbox:yangge1987@gmail.com": {
-        access_token: "fill in"
-    },
-    "github:jam-world": {
-        access_token: "fill in here"
-    }
+    "dropbox:yangge1987@gmail.com": dropboxAccountHelper(),
+    "github:jam-world": githubAccountHelper()
```
Comment (episodeyang) [inline file-browser/src/model/model.js]: 1. [ ] call `githubAccountHelper` -> `githubAccountKey`. 
2. [ ] key helper need to have the signature 
```jsx
type KeyHelper (service:string, username?:string, email?:string)=> string;
```

File: file-browser/src/services/dropbox.js [hunk 10]
```diff
@@ -27,6 +27,10 @@ import {
     fork
 } from "luna-saga";
 
+// test only
+const testClientId : string = "uhpeu63dv1pqgor";
+const testAccessToken : string = "p5mHBr4rHvYAAAAAAAAANpuR45x2fOWdBSQqs5rtKeFD8GMF7uH8emUZTy_F53zT";
```
Comment (episodeyang) [inline file-browser/src/services/dropbox.js]: put this inside the model/store file. Keep dropbox.js pure.

File: file-browser/src/services/dropbox.js [hunk 11]
```diff
@@ -35,3 +39,20 @@ export function* dropboxMainProc() {
     // fork(new_process, arguments1, arg2, arg3...);
     // }
 }
+
+
+const uuidV4 = require('uuid/v4');
+export function accountHelper(cliendId = testClientId, accessToken = testAccessToken, uuidGenerator = uuidV4) {
```
Comment (episodeyang) [inline file-browser/src/services/dropbox.js]: this shouldn't been account helper. 

- [ ] we need: `accountKey` helper

uuidGenerator shouldn't be there either.

File: file-browser/package.json (hunk 2 above)
Comment (episodeyang) [inline file-browser/package.json]: I got this one. See comment below, the config session might need to be reviewed.

File: file-browser/package.json (hunk 3 above)
Comment (jam-world) [inline file-browser/package.json]: Not necessary. NPM module is included by accident, when we are fixing the npm itself.

File: file-browser/package.json (hunk 4 above)
Comment (jam-world) [inline file-browser/package.json]: Done.

File: file-browser/package.json (hunk 5 above)
Comment (jam-world) [inline file-browser/package.json]: Done.

File: file-browser/package.json (hunk 3 above)
Comment (episodeyang) [inline file-browser/package.json]: - [ ] remove npm from package.json @jam-world

File: file-browser/public/electron.js (hunk 7 above)
Comment (jam-world) [inline file-browser/public/electron.js]: Modified to win.loadURL(isDev ? process.env.ELECTRON_START_URL : `file://${path.join(__dirname, '../build/index.html')}`);

Comment (episodeyang) [inline file-browser/public/electron.js]: @jam-world can you check the checkbox *after* you push a commit? Thanks :hugs:

File: file-browser/src/model/dataSchema.js (hunk 8 above)
Comment (episodeyang) [inline file-browser/src/model/dataSchema.js]: Let's leave these as they are. We can come back to it tonight during our big refactor.

File: file-browser/public/electron.js [hunk 12]
```diff
@@ -9,14 +10,8 @@ function createWindow() {
...
-    // and load the index.html of the app.
-    // from https://medium.freecodecamp.com/building-an-electron-application-with-create-react-app-97945861647c
-    const startUrl = process.env.ELECTRON_START_URL || url.format({
-            pathname: path.join(__dirname, 'build/index.html'),
-            protocol: 'file:',
-            slashes: true
-        });
-    win.loadURL(startUrl);
+    // load the url
+    win.loadURL(isDev ? process.env.ELECTRON_START_URL : `file://${path.join(__dirname, '../build/index.html')}`);
```
Comment (episodeyang) [inline file-browser/public/electron.js]: This still seems wrong. Did you test this?

Comment (jam-world) [inline file-browser/public/electron.js]: Yes, I have tested it.

**Comment by jam-world** [comment]
The github token is outdated again. You may need to get a new token for the test purpose.

File: file-browser/public/electron.js (hunk 12 above)
Comment (episodeyang) [inline file-browser/public/electron.js]: So no default URL huh 🤓

**Comment by episodeyang** [comment]
Thanks @jam-world ! I will go ahead and merge this

**Comment by jam-world** [comment]
I haven't finish the web publish pipeline. However, I have moved all the test information to the model module, You can play with it now.

**Comment by episodeyang** [comment]
Thanks. I will clean up the model code
